flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/agglomerativeclustering/AgglomerativeClustering.java

```diff
@@ -0,0 +1,415 @@
+                                            distanceMeasure.distance(
+                                                    vectorWithNorm1, vectorWithNorm2));
+                        }
+                    }
+                    break;
+                case LINKAGE_SINGLE:
+                    distance = Double.MAX_VALUE;
+                    for (int i = 0; i < size1; i++) {
+                        for (int j = 0; j < size2; j++) {
+                            VectorWithNorm vectorWithNorm1 =
+                                    vectorWithNorms[cluster1.dataPointIds.get(i)];
+                            VectorWithNorm vectorWithNorm2 =
+                                    vectorWithNorms[cluster2.dataPointIds.get(j)];
+                            distance =
+                                    Math.min(
+                                            distance,
+                                            distanceMeasure.distance(
+                                                    vectorWithNorm1, vectorWithNorm2));
+                        }
+                    }
+                    break;
+                case LINKAGE_WARD:
+                    int vecSize = vectorWithNorms[0].vector.size();
+                    DenseVector mean1 = Vectors.dense(new double[vecSize]);
+                    DenseVector mean2 = Vectors.dense(new double[vecSize]);
+
+                    for (int i = 0; i < size1; i++) {
+                        BLAS.axpy(1.0, vectorWithNorms[cluster1.dataPointIds.get(i)].vector, mean1);
+                    }
+                    for (int i = 0; i < size2; i++) {
+                        BLAS.axpy(1.0, vectorWithNorms[cluster2.dataPointIds.get(i)].vector, mean2);
+                    }
+
+                    DenseVector meanMerged = mean1.clone();
+                    BLAS.axpy(1.0, mean2, meanMerged);
+                    BLAS.scal(1.0 / size1, mean1);
+                    BLAS.scal(1.0 / size2, mean2);
+                    BLAS.scal(1.0 / (size1 + size2), meanMerged);
+                    double essInc =
+                            size1 * BLAS.dot(mean1, mean1)
+                                    + size2 * BLAS.dot(mean2, mean2)
+                                    - (size1 + size2) * BLAS.dot(meanMerged, meanMerged);
+
+                    distance = Math.sqrt(2 * essInc);
+                    break;
+                default:
+                    throw new UnsupportedOperationException(
+                            "Unsupported " + LINKAGE + " type: " + linkage + ".");
+            }
+            return distance;
+        }
+
+        /** A cluster with cluster Id specified and data points that belong to this cluster. */
+        private static class Cluster {
+            private final int clusterId;
+            private final List<Integer> dataPointIds;
+
+            @SafeVarargs
+            public Cluster(
+                    int clusterId, List<Integer> dataPointIds, List<Integer>... otherDataPointIds) {
```

Comment:
It might be better to write two constructors, `Cluster(int, List<Integer>)` and `Cluster(int, List<Integer>, List<Integer>)`.